Sources/Core/Shared/GroupOperation.swift

```diff
@@ -339,3 +448,117 @@ public struct WillAddChildObserver: GroupOperationWillAddChildObserver {
         didAttachToOperation?(operation: operation)
     }
 }
+
+private extension GroupOperation {
+    /**
+     The group operation handles thread-safe addition of operations by utilizing two final operations:
+     - a CanFinishOperation which manages handling GroupOperation internal state and has every child
+       operation as a dependency
+     - a finishingOperation, which has the CanFinishOperation as a dependency
+
+     The purpose of this is to handle the possibility that GroupOperation.addOperation() or
+     GroupOperation.queue.addOperation() are called right after all current child operations have
+     completed (i.e. after the CanFinishOperation has been set to ready), but *prior* to being able
+     to process that the GroupOperation is finishing (i.e. prior to the CanFinishOperation executing and
+     acquiring the GroupOperation.groupFinishLock to set state).
+     */
+    private class CanFinishOperation: NSOperation {
+        private weak var parent: GroupOperation?
+        private var _finished = false
+
+        init(parentGroupOperation: GroupOperation) {
+            self.parent = parentGroupOperation
+            super.init()
+        }
+        override func start() {
+            main()
+        }
+        override func main() {
+            execute()
+        }
+        func execute() {
+            if let parent = parent {
+
+                // All operations that were added as a side-effect of anything up to
+                // WillFinishObservers of prior operations should have been executed.
+                //
+                // Handle an edge case caused by concurrent calls to GroupOperation.addOperations()
+
+                let isWaiting = parent.groupFinishLock.withCriticalScope { () -> Bool in
+
+                    // Is anything currently adding operations?
+                    guard dispatch_group_wait(parent.isAddingOperationsGroup, DISPATCH_TIME_NOW) == 0 else {
+                        // Operations are actively being added to the group
+                        // Wait for this to complete before proceeding.
+                        dispatch_group_notify(parent.isAddingOperationsGroup,
+                            dispatch_get_global_queue(NSQualityOfServiceToDispatchQOS(self.qualityOfService), 0),
+                            { [unowned self] in
+                            self.execute()
+                        })
```

Comment:
I think it would be great to add an initialiser to `Queue` here which takes a `qos_class` and/or `NSQualityOfService`. See Queue.swift.

```swift
    /**
     Initialize a Queue with a given NSQualityOfService.
     
     - parameter qos: a NSQualityOfService value
     - returns: a Queue with an equivalent quality of service
    */
    public init(qos: NSQualityOfService) {
        switch qos {
        case .Background:
            self = .Background
        case .Default:
            self = .Default
        case .UserInitiated:
            self = .Initiated
        case .UserInteractive:
            self = .Interactive
        case .Utility:
            self = .Utility
        }
    }

    /**
     Initialize a Queue with a given GCD quality of service class.

     - parameter qos: a qos_class_t value
     - returns: a Queue with an equivalent quality of service
     */
    public init(qos: qos_class_t) {
        switch qos {
        case qos_class_main():
            self = .Main
        case QOS_CLASS_BACKGROUND:
            self = .Background
        case QOS_CLASS_USER_INITIATED:
            self = .Initiated
        case QOS_CLASS_USER_INTERACTIVE:
            self = .Interactive
        case QOS_CLASS_UTILITY:
            self = .Utility
        default:
            self = .Default
        }
    }
```

If you add that into Queue.swift, then your code above can become:

```swift
dispatch_group_notify(parent.isAddingOperationsGroup, Queue(qos: qualityOfService).queue, { [unowned self] in
    self.execute()
})
```

Also, is it not possible to just pass `execute` directly?

```swift
dispatch_group_notify(parent.isAddingOperationsGroup, Queue(qos: qualityOfService).queue, execute)
```